hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java

```diff
@@ -2483,8 +2483,15 @@ public boolean areReleaseResourceEnabled() {
   /**
    * Returns whether the explicit guard of lock is required.
    */
-  public boolean needsLockGuard() {
-    return isMetadataTableEnabled() || getWriteConcurrencyMode().supportsOptimisticConcurrencyControl();
+  public boolean isLockRequired() {
+    return !isDefaultLockProvider() || getWriteConcurrencyMode().supportsOptimisticConcurrencyControl();
```

Comment:
Default lock provider is zk lock.
i guess that when MDT enabled,  **autoAdjustConfigsForConcurrencyMode**() will  set  InProcessLockProvider to LockProvider,   so we use isDefaultLockProvider() instead of isMetadataTableEnabled()

Reply:
Yes, to be compatible with the exising behavior, if the lock provider is reset by auto adjust, then it means it is definite that the lock is required.

MDT enabled does not mean we need a lock provider(check the premise of auto adjust lock actions), the change in #8111 is a little aggresive, so fix it altogether.

Reply:
The user can still set ZK lock for a single writer with sync table services and this method returns false in this case, which is unexpected.

Reply:
Yeah, I was expecting the use to set up the optimistic concurrency control explicitly.